src/components/InviteFriends/index.tsx

```diff
@@ -0,0 +1,220 @@
+import { useEffect, useMemo, useState } from "react";
+import AElf from "aelf-sdk";
+import clsx from "clsx";
+import Drawer from "../Drawer";
+import { InviteDetail, IStartAppParams, ShortLinkRes } from "@/types/task";
+import { stringifyStartAppParams } from "@/utils/start-params";
+import { connectUrl, portkeyServer, TgLink } from "@/config";
+import { useWalletService } from "@/hooks/useWallet";
+import { getCaHashAndOriginChainIdByWallet } from "@/utils/wallet";
+import { chainId } from "@/constants/app";
+import { useRequest } from "ahooks";
+import Loading from "../Loading";
+import { useCopyToClipboard } from "react-use";
+import { QRCode } from "react-qrcode-logo";
+
+interface IInviteFriendsStatusProps {
+  data?: InviteDetail;
+  isShowDrawer?: boolean;
+  onClickInvite?(): void;
+}
+
+const InviteFriendsStatus = ({
+  data,
+  isShowDrawer,
+  onClickInvite,
+}: IInviteFriendsStatusProps) => {
+  const [, setCopied] = useCopyToClipboard();
+  const [isCopied, setIsCopied] = useState(false);
+  const { wallet, walletType } = useWalletService();
+  const progress = useMemo(() => {
+    if (!data?.totalInvitesNeeded) return 0;
+    const percentage =
+      ((data?.votigramVoteAll || 0) / data?.totalInvitesNeeded) * 100;
+    return percentage < 100 ? percentage : 100;
+  }, [data]);
+
+  const generateCode = async () => {
+    const timestamp = Date.now();
+    const didWallet = wallet?.extraInfo?.portkeyInfo;
+    const message = Buffer.from(
+      `${didWallet?.walletInfo.address}-${timestamp}`
+    ).toString("hex");
+    const signature = AElf.wallet
+      .sign(message, didWallet?.walletInfo.keyPair)
+      .toString("hex");
+    const publicKey = wallet?.extraInfo?.publicKey ?? "";
+    const { caHash } = await getCaHashAndOriginChainIdByWallet(
+      wallet,
+      walletType
+    );
+    const requestObject = {
+      grant_type: "signature",
+      client_id: "CAServer_App",
+      scope: "CAServer",
+      signature: signature,
+      pubkey: publicKey,
+      timestamp: timestamp.toString(),
+      ca_hash: caHash,
+      chain_id: chainId,
+    };
+    const portKeyRes = await fetch(connectUrl + "/connect/token", {
+      method: "POST",
+      body: new URLSearchParams(requestObject).toString(),
+      headers: {
+        "Content-Type": "application/x-www-form-urlencoded",
+      },
+    });
+    const portKeyResObj = await portKeyRes.json();
+    if (portKeyRes.ok) {
+      const token = portKeyResObj.access_token;
```

Comment:
what if access_token is undefined? do you still proceed?

Reply:
Updated